File: src/main/java/me/miki/shindo/management/account/AccountManager.java
package me.miki.shindo.management.account;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import me.miki.shindo.Shindo;
import me.miki.shindo.libs.openauth.microsoft.MicrosoftAuthenticationException;
import me.miki.shindo.injection.interfaces.IMixinMinecraft;
import me.miki.shindo.logger.ShindoLogger;
import me.miki.shindo.management.file.FileManager;
import me.miki.shindo.utils.JsonUtils;
import me.miki.shindo.utils.Multithreading;
import me.miki.shindo.utils.SessionUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.Session;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.UUID;

public class AccountManager {

    private final ArrayList<Account> accounts = new ArrayList<Account>();
    private String currentAccount;

    public AccountManager() {

        FileManager fileManager = Shindo.getInstance().getFileManager();
        File microsoftDir = new File(fileManager.getExternalDir(), "microsoft");
        File accountFile = new File(fileManager.getShindoDir(), "Account.json");
        File skinDir = new File(fileManager.getCacheDir(), "skin");

        if (!microsoftDir.exists()) {
            fileManager.createDir(microsoftDir);
        }

        if (!accountFile.exists()) {
            fileManager.createFile(accountFile);
        }

        if (!skinDir.exists()) {
            fileManager.createDir(skinDir);
        }

        if (accountFile.length() > 0) {
            load();
        }

        Account storedAccount = getAccountByName(currentAccount);
        if (storedAccount != null) {

            if (storedAccount.getType().equals(AccountType.MICROSOFT)) {
                ShindoLogger.info("Login into Microsoft account");
                Account acc = storedAccount;
                if (acc.getRefreshToken() == null || acc.getRefreshToken().isEmpty()) {
                    ShindoLogger.warn("Stored Microsoft account is missing a refresh token. Please sign in again.");
                } else {
                    Multithreading.runAsync(() -> {
                        try {
                            SessionUtils.getInstance().loginMicrosoftAccount(acc);
                            save();
                        } catch (MicrosoftAuthenticationException e) {
                            ShindoLogger.error("Failed to login Microsoft account", e);
                        }
                    });
                }
            } else {
                ShindoLogger.info("Login into Offline account");
                Account acc = storedAccount;
                File f = new File(skinDir, acc.getName() + ".png");

                UUID offlineId = UUID.nameUUIDFromBytes(
                        ("OfflinePlayer:" + acc.getName()).getBytes(StandardCharsets.UTF_8)
                );

                Minecraft mc = Minecraft.getMinecraft();
                ((IMixinMinecraft) mc).setSession(new Session(acc.getName(), offlineId.toString(), "0", "mojang"));

                if (f.exists()) {
                    acc.setSkinFile(f);
                } else {
                    mc.getTextureManager().bindTexture(new ResourceLocation("textures/entity/steve.png"));
                }
            }
        }
    }

    public void save() {

        FileManager fileManager = Shindo.getInstance().getFileManager();

        try (FileWriter writer = new FileWriter(new File(fileManager.getShindoDir(), "Account.json"))) {

            Gson gson = new Gson();
            JsonObject jsonObject = new JsonObject();
            JsonArray jsonArray = new JsonArray();

            jsonObject.addProperty("Current Account", currentAccount);

            for (Account acc : accounts) {

                JsonObject accJsonObject = new JsonObject();

                accJsonObject.addProperty("Name", acc.getName());
                accJsonObject.addProperty("UUID", acc.getUuid());
                accJsonObject.addProperty("Account Type", acc.getType().getId());
                accJsonObject.addProperty("RefreshToken", acc.getRefreshToken());

                jsonArray.add(accJsonObject);
            }

            jsonObject.add("Accounts", jsonArray);

            gson.toJson(jsonObject, writer);
        } catch (Exception e) {
            ShindoLogger.error("Failed to save account", e);
        }
    }

    public void load() {

        FileManager fileManager = Shindo.getInstance().getFileManager();

        try (FileReader reader = new FileReader(new File(fileManager.getShindoDir(), "Account.json"))) {

            Gson gson = new Gson();
            JsonObject jsonObject = gson.fromJson(reader, JsonObject.class);

            if (jsonObject != null && jsonObject.isJsonObject()) {

                JsonArray jsonArray = JsonUtils.getArrayProperty(jsonObject, "Accounts");

                currentAccount = JsonUtils.getStringProperty(jsonObject, "Current Account", "null");

                if (jsonArray != null) {

                    Iterator<JsonElement> iterator = jsonArray.iterator();

                    while (iterator.hasNext()) {

                        JsonElement jsonElement = iterator.next();
                        JsonObject accJsonObject = gson.fromJson(jsonElement, JsonObject.class);

                        Account account = new Account(
                                JsonUtils.getStringProperty(accJsonObject, "Name", "null"),
                                JsonUtils.getStringProperty(accJsonObject, "UUID", "null"),
                                AccountType.getAccountTypeById(JsonUtils.getIntProperty(accJsonObject, "Account Type", 0)
                                )
                        );
                        account.setRefreshToken(JsonUtils.getStringProperty(accJsonObject, "RefreshToken", ""));
                        accounts.add(account);
                    }
                }
            }
        } catch (Exception e) {
            ShindoLogger.error("Failed to load account", e);
        }
    }

    public Account getCurrentAccount() {
        return getAccountByName(currentAccount);
    }

    public void setCurrentAccount(Account account) {
        this.currentAccount = account.getName();
    }

    public Account getAccountByName(String name) {

        for (Account acc : accounts) {
            if (acc.getName().equals(name)) {
                return acc;
            }
        }

        return null;
    }

    public Account getAccountByUuid(String uuid) {
        if (uuid == null || uuid.equalsIgnoreCase("null")) {
            return null;
        }

        for (Account acc : accounts) {
            if (acc.getUuid() != null && acc.getUuid().equalsIgnoreCase(uuid)) {
                return acc;
            }
        }

        return null;
    }

    public ArrayList<Account> getAccounts() {
        return accounts;
    }

}

File: src/main/kotlin/me/miki/shindo/management/account/AccountManager.kt
package me.miki.shindo.management.account

import com.google.gson.JsonParser
import me.miki.shindo.api.websocket.AccountType
import me.miki.shindo.logger.ShindoLogger
import me.miki.shindo.management.account.data.SavedAccount
import me.miki.shindo.management.file.FileManager
import net.minecraft.client.Minecraft
import net.minecraft.util.Session
import net.raphimc.minecraftauth.MinecraftAuth
import net.raphimc.minecraftauth.java.JavaAuthManager
import net.raphimc.minecraftauth.msa.model.MsaDeviceCode
import net.raphimc.minecraftauth.msa.service.impl.DeviceCodeMsaAuthService
import net.raphimc.minecraftauth.msa.service.util.ParamMsaAuthServiceSupplier
import java.lang.reflect.Field
import java.util.UUID
import java.util.function.Consumer

class AccountManager(
    fileManager: FileManager,
) {
    private val storage = AccountStorage(fileManager)
    private val accounts: MutableList<SavedAccount> = storage.load().toMutableList()

    fun getAccounts(): List<SavedAccount> = accounts.toList()

    fun getActiveAccount(): SavedAccount? = accounts.firstOrNull { it.active }

    fun switchAccount(id: String) {
        val target = accounts.firstOrNull { it.id == id } ?: return
        val updated = accounts.map { it.copy(active = it.id == id) }
        accounts.clear()
        accounts.addAll(updated)
        save()
        injectSession(target)
    }

    fun addMicrosoftAccount(
        onCode: (DeviceCodeInfo) -> Unit,
        onSuccess: (SavedAccount) -> Unit,
        onFailure: (Exception) -> Unit,
    ) {
        try {
            val httpClient = MinecraftAuth.createHttpClient()

            val authManager =
                JavaAuthManager.create(httpClient).login(
                    ParamMsaAuthServiceSupplier { client, appConfig, param ->
                        DeviceCodeMsaAuthService(client, appConfig, param)
                    },
                    Consumer<MsaDeviceCode> { code ->
                        onCode(
                            DeviceCodeInfo(
                                userCode = code.userCode,
                                verificationUrl = code.verificationUri,
                                directUrl = code.directVerificationUri,
                                expiresIn = code.expireTimeMs,
                            ),
                        )
                    },
                )

            // Eagerly request tokens so they are all included in the serialized chain
            val token = authManager.minecraftToken.upToDate
            val profile = authManager.minecraftProfile.upToDate

            val acc =
                SavedAccount(
                    id = UUID.randomUUID().toString(),
                    type = AccountType.MICROSOFT,
                    username = profile.name,
                    uuid = profile.id.toString().replace("-", ""),
                    accessToken = token.token,
                    tokenExpiresAt = token.expireTimeMs,
                    sessionJson = JavaAuthManager.toJson(authManager).toString(),
                    active = false,
                )

            // Persist whenever the library refreshes tokens in the background
            authManager.changeListeners.add {
                updateSessionJson(acc.id, JavaAuthManager.toJson(authManager).toString())
            }

            addOrReplace(acc)
            onSuccess(acc)
        } catch (e: Exception) {
            ShindoLogger.error("Microsoft auth failed", e)
            onFailure(e)
        }
    }

    fun addOfflineAccount(username: String): SavedAccount {
        val uuid =
            UUID
                .nameUUIDFromBytes(
                    "OfflinePlayer:$username".toByteArray(Charsets.UTF_8),
                ).toString()
                .replace("-", "")

        val acc =
            SavedAccount(
                id = UUID.randomUUID().toString(),
                type = AccountType.OFFLINE,
                username = username,
                uuid = uuid,
                accessToken = "-",
                tokenExpiresAt = 0L,
                sessionJson = "",
                active = false,
            )
        addOrReplace(acc)
        return acc
    }

    fun refreshIfNeeded(id: String): SavedAccount {
        val acc =
            accounts.firstOrNull { it.id == id }
                ?: throw IllegalArgumentException("Account $id not found")

        if (acc.type == AccountType.OFFLINE || acc.sessionJson.isEmpty()) return acc

        val fiveMinutes = 5L * 60L * 1000L
        if (System.currentTimeMillis() < acc.tokenExpiresAt - fiveMinutes) return acc

        return try {
            val httpClient = MinecraftAuth.createHttpClient()
            val authManager =
                JavaAuthManager.fromJson(
                    httpClient,
                    JsonParser.parseString(acc.sessionJson).asJsonObject,
                )

            val token = authManager.minecraftToken.upToDate
            val profile = authManager.minecraftProfile.upToDate

            val updated =
                acc.copy(
                    username = profile.name,
                    accessToken = token.token,
                    tokenExpiresAt = token.expireTimeMs,
                    sessionJson = JavaAuthManager.toJson(authManager).toString(),
                )
            replaceById(updated)
            save()
            updated
        } catch (e: Exception) {
            ShindoLogger.error("Token refresh failed for ${acc.username} — user may need to re-login", e)
            acc
        }
    }

    fun removeAccount(id: String) {
        val wasActive = accounts.firstOrNull { it.id == id }?.active == true
        accounts.removeAll { it.id == id }
        if (wasActive && accounts.isNotEmpty()) {
            val first = accounts[0].copy(active = true)
            accounts[0] = first
            injectSession(first)
        }
        save()
    }

    fun injectSession(account: SavedAccount) {
        try {
            val mc = Minecraft.getMinecraft()
            val token = if (account.type == AccountType.OFFLINE) "-" else account.accessToken
            val session = Session(account.username, account.uuid, token, "mojang")
            val field: Field = Minecraft::class.java.getDeclaredField("session")
            field.isAccessible = true
            field.set(mc, session)
            ShindoLogger.info("Session injected for ${account.username} (${account.type})")
        } catch (e: Exception) {
            ShindoLogger.error("Failed to inject session for ${account.username}", e)
        }
    }

    private fun addOrReplace(acc: SavedAccount) {
        // If the same MS/offline account is added again, replace instead of duplicating
        val idx = accounts.indexOfFirst { it.uuid == acc.uuid && it.type == acc.type }
        if (idx >= 0) accounts[idx] = acc else accounts.add(acc)
        save()
    }

    private fun replaceById(updated: SavedAccount) {
        val idx = accounts.indexOfFirst { it.id == updated.id }
        if (idx >= 0) accounts[idx] = updated
    }

    private fun updateSessionJson(
        id: String,
        json: String,
    ) {
        val idx = accounts.indexOfFirst { it.id == id }
        if (idx >= 0) {
            accounts[idx] = accounts[idx].copy(sessionJson = json)
            save()
        }
    }

    private fun save() = storage.save(accounts)

    data class DeviceCodeInfo(
        val userCode: String,
        val verificationUrl: String,
        val directUrl: String,
        val expiresIn: Long,
    )
}
